test/snippet/alignment/matrix/debug_matrix.cpp

```diff
@@ -0,0 +1,121 @@
+
+debug_matrix score_matrix
+{
+    std::vector
+    {
+        -0, -1, -2, -3, -4, -5, -6, -7, -8,
+        -1, -0, -1, -2, -3, -4, -5, -6, -7,
+        -2, -1, -1, -1, -2, -3, -4, -5, -6,
+        -3, -2, -2, -2, -2, -2, -3, -4, -5,
+        -4, -3, -3, -3, -3, -3, -3, -3, -4
+    },
+    5u,
+    9u
+};
+
+debug_stream << "database:\t" << (database | view::to_char) << std::endl;
+debug_stream << "query:\t\t" << (query | view::to_char) << std::endl;
+debug_stream << std::endl;
+
+debug_stream << "score_matrix: " << score_matrix.cols() << " columns and "
+          << score_matrix.rows() << " rows" << std::endl;
+
+// Print out the matrix.
+for(unsigned row = 0u; row < score_matrix.rows(); ++row)
+{
+    for(unsigned col = 0u; col < score_matrix.cols(); ++col)
+        debug_stream << score_matrix.at(row, col) << ", ";
+    debug_stream << std::endl;
+}
+debug_stream << std::endl;
+
+// Prints out the matrix in a convenient way
+debug_stream << score_matrix << std::endl; // without sequences
+debug_stream << debug_matrix{score_matrix, database, query} << std::endl; // with sequences
+debug_stream << fmtflags2::utf8 << debug_matrix{score_matrix, database, query}; // as utf8
+//! [score_matrix]
+    debug_stream << "//! [score_matrix::out]" << std::endl;
+    debug_stream.unsetf(fmtflags2::utf8);
+}
+{
+    debug_stream << "//! [trace_matrix::out]" << std::endl;
+//! [trace_matrix]
+using namespace seqan3;
+using namespace seqan3::detail;
+
+std::vector<dna4> database = "AACCGGTT"_dna4;
+std::vector<dna4> query = "ACGT"_dna4;
+
+auto N = trace_directions::none;
+auto D = trace_directions::diagonal;
+auto U = trace_directions::up;
+auto L = trace_directions::left;
+
+debug_matrix trace_matrix
+{
+    std::vector
+    {
+        N,L,L  ,L  ,L  ,L  ,L  ,L,L  ,
+        U,D,D|L,L  ,L  ,L  ,L  ,L,L  ,
+        U,U,D  ,D  ,D|L,L  ,L  ,L,L  ,
+        U,U,D|U,D|U,D  ,D  ,D|L,L,L  ,
+        U,U,D|U,D|U,D|U,D|U,D  ,D,D|L
+    },
+    5u,
+    9u
+};
+
+debug_stream << "database:\t" << (database | view::to_char) << std::endl;
+debug_stream << "query:\t\t" << (query | view::to_char) << std::endl;
+debug_stream << std::endl;
+
+debug_stream << "trace_matrix: " << trace_matrix.cols() << " columns and "
+          << trace_matrix.rows() << " rows" << std::endl;
+
+// Print out the matrix.
+for(unsigned row = 0u; row < trace_matrix.rows(); ++row)
+{
+    for(unsigned col = 0u; col < trace_matrix.cols(); ++col)
+    {
+        trace_directions dir = trace_matrix.at(row, col);
+        if (dir == N)
+            debug_stream << "N";
+        if ((dir & D) == D)
+            debug_stream << "D";
+        if ((dir & U) == U)
+            debug_stream << "U";
+        if ((dir & L) == L)
+            debug_stream << "L";
+        debug_stream << ", ";
+    }
+    debug_stream << std::endl;
+}
+debug_stream << std::endl;
+
+// Prints out the matrix in a convenient way
+debug_stream << trace_matrix << std::endl; // without sequences
```

Comment:
why do you show how to "not conveniently" print a matrix above?  I think the two for loops are trivial and what you actually want to show is the convenient way. Or am I wrong?

Reply:
I don't understand your question. `debug_stream << trace_matrix << std::endl;` prints the matrix directly.

The above code shows how to do it manually.

Should I remove the manual example?

Reply:
> Should I remove the manual example?

IMHO yes. I don't see the benefit of showing people how to iterate over a matrix?  
And I am always in favor of only showing "how to do it the right way" and omitting the "wrong" way completely. But if you have an argument to leave it I don't have strong feelings about this.